### PR TITLE
Highlight mentions in chat input

### LINKED ISSUE #16097: Message - Composer - Mentions not resolving in editor
Currently the mentions show no clear UI to the user that the mention has been correctly resolved.

Composer_for_Mobile_Figma_2023-05-30_16-41-12

See the designs for how this should look:
https://www.figma.com/file/wA8Epdki2OWa8Vr067PCNQ/Composer-for-Mobile?type=design&node-id=5530-358489&t=1xpxUi8twtiGgxdm-0

### PR DESCRIPTION
**Pull request by vkjr**
fixes #16097

### Summary

Highlight mentions in user input 

Screenshot 2023-06-20 at 14 08 12

## Reviewer notes
If you have any idea on how to overcome the issue described in the next section, please tell me. 
I also tried to implement similar mentions in the test react-native app and the similar bug is there, but barely noticeable because the basic test app is much faster. 
It is not a bug in `status-go` mentions implementation, because we are getting correct structure from there.
And it is not a bug in generated children Text components, they are also correct.
So I believe it is a `react-native` bug.

## Testing notes
There is a known issue, steps to reproduce:
- enter some mention
- remove whitespace after mention, put the coma
- start quick typing

Result: all typed characters get the style of the mention for a fraction of a second. If you remove some characters and start typing again - the effect disappears. On the physical iPhone 13 this is barely visible. On the Android simulator, it is very visible (but it is generally terribly slow)

Video:

https://github.com/status-im/status-mobile/assets/5786310/811f729d-8563-4a4b-8246-6cc0a3204577

I'm not sure how critical the issue is on physical Android.

status: ready

### CODE REVIEW

File: src/status_im2/contexts/chat/composer/view.cljs
Comment (cammellos) [inline src/status_im2/contexts/chat/composer/view.cljs]:
```
(condp = chunk-type
   :text  (conj..
   :mention (conj ...)
```
Avoids `case` as it's not safe and it's a bit more compact than just cond

Comment (vkjr) [reply src/status_im2/contexts/chat/composer/view.cljs]: Thanks!

File: src/status_im2/contexts/chat/composer/effects.cljs
Comment (J-Son89) [inline src/status_im2/contexts/chat/composer/effects.cljs]: probably you should include a TODO here with a link for the issue to restore this functionality 👍

File: src/status_im2/contexts/chat/composer/style.cljs
Comment (J-Son89) [inline src/status_im2/contexts/chat/composer/style.cljs:134]: consider changing key to `:suppress-highlighting?`

File: src/status_im2/contexts/chat/composer/view.cljs
Comment (J-Son89) [inline src/status_im2/contexts/chat/composer/view.cljs:28]: extra new line

File: src/status_im2/contexts/chat/composer/view.cljs
Comment (J-Son89) [inline src/status_im2/contexts/chat/composer/view.cljs]: maybe same TODO here?

Comment (vkjr) [reply src/status_im2/contexts/chat/composer/view.cljs]: @J-Son89, it is "wip" PR, created for the sake of getting ios builds for device) Don't need a review yet, since it is not a final version, there is still a bug that I'm trying to solve) All code pieces that are not needed will be removed of course, not commented)

Comment (J-Son89) [reply src/status_im2/contexts/chat/composer/view.cljs]: Apologies, I missed that. Please ignore my comments so 😅

**Review by J-Son89 — APPROVED** [review]
nice @vkjr!

File: src/status_im2/contexts/chat/composer/handlers.cljs
Comment (OmarBasem) [inline src/status_im2/contexts/chat/composer/handlers.cljs:127]: Why this change? I think this is to avoid many calls to status-go for mentions cc @qfrank

**Review by OmarBasem — APPROVED** [review]

File: src/status_im2/contexts/chat/composer/handlers.cljs
Comment (qfrank) [reply src/status_im2/contexts/chat/composer/handlers.cljs:127]: Also would like to know why 🙂

Comment (vkjr) [reply src/status_im2/contexts/chat/composer/handlers.cljs:127]: @OmarBasem, @qfrank, two things:

1) we need to get updated text structure from `status-go` as quick as possible, because while it is not updated, highlighting is not applied

2) With previous logic I experienced a bug. Imagine that user is going to type "hello @". The order of event is:
- user types "hello"
- user types whitespace after "hello", `on-change-text` scheduled with delay and with content "hello "
- user types "@", `on-change-text` immediately processed and current text input becomes "hello @"
- scheduled `on-change-text` processed with delay and current text get backs to the "hello ", which results in losing the "@"

Comment (qfrank) [reply src/status_im2/contexts/chat/composer/handlers.cljs:127]: emm.. what if we enhance `debounce/debounce-and-dispatch` to make it support dispatch immediately if user input end with @ ?  @vkjr

Comment (vkjr) [reply src/status_im2/contexts/chat/composer/handlers.cljs:127]: @qfrank, I think `debounce-and-dispatch` is too general to make this change, because it works not only with `on-change-text`, but even if we add this hack, that won't eliminate the problem with slow highlight updating.
Imagine that user typed `@asd` and continued quick typing. All the symbols after "@" won't be highlighted because while user quickly types, events not sent and updated text structure is not received from `status-go`

Comment (qfrank) [reply src/status_im2/contexts/chat/composer/handlers.cljs:127]: yeah, quick typing would make weired behavior. 😅

> debounce-and-dispatch is too general 

what if make a new function like `debounce-and-dispatch` and add an extra trigger condition to the actual invoking e.g. the number of user input characters reached a specific value. just my thoughts , it's up to you, if we found performance issue with mention, there still exist spaces to improve in status-go side 🙂

Comment (vkjr) [reply src/status_im2/contexts/chat/composer/handlers.cljs:127]: @qfrank, I think right now it is more a preliminary optimization) Let's see if there are any issues with existing implementation and then start improving, wdyt? I'm currently more concerned about that bug with highlighting that is in the description. I believe it can be a showstopper on slow devices :-/

Comment (qfrank) [reply src/status_im2/contexts/chat/composer/handlers.cljs:127]: agree

**Review by siddarthkay — APPROVED** [review]

**Comment by pavloburykh** [comment]
@vkjr hi! Could you please clarify if recent commits is just rebasing or some additional changes? Asking case currently I am testing builds from 2023-06-27. As there is IOS [blocker](https://github.com/status-im/status-mobile/issues/16421) in develop I will not be able to test recent rebased PR build as it it will contain this bug.

**Comment by vkjr** [comment]
@pavloburykh, there is no additional functionality, I just rebased to the latest develop before figuring out there is a blocker :-/ Testing yesterday's build is fine, no changes since then)

**Comment by pavloburykh** [comment]
> @pavloburykh, there is no additional functionality, I just rebased to the latest develop before figuring out there is a blocker :-/ Testing yesterday's build is fine, no changes since then)

Thank you @vkjr !

**Comment by pavloburykh** [comment]
@vkjr thanx for the PR! Please take a look at the following issues

### ISSUE 1 Erratic Cursor Behavior and Missing Characters During Text Entry and Deletion (Android)

Reproduced on Android 12, Samsung Galaxy A52

**Steps:** 

1. Open any chat
2. Start entering some text into input field
3. Observe cursor's behaviour

**Actual result:** the cursor exhibits unpredictable behavior by jumping from symbol to symbol instead of maintaining a steady position. This erratic cursor movement leads to instances where the first letter of a word entered by the user simply vanishes, causing partial or incomplete words. Additionally, when attempting to delete text, the cursor sometimes skips over certain symbols, preventing their deletion.


https://github.com/status-im/status-mobile/assets/97245802/28538512-73e8-420c-a9c9-7f2798c750ac

**Comment by vkjr** [comment]
@pavloburykh, thanks! Do I understand correctly that on iOS devices such behavior doesn't happen?

**Comment by pavloburykh** [comment]
> @pavloburykh, thanks! Do I understand correctly that on iOS devices such behavior doesn't happen?

@vkjr correct. Android only.

**Comment by vkjr** [comment]
@pavloburykh, could you please check the cursor fix on Android? I don't have a physical device, but on the Simulator it works better)

**Comment by du82** [comment]
@vkjr it still happens for me on Android

**Comment by vkjr** [comment]
@ du82, could you upload a video please? Does the cursor jump as much as in the original issue?

**Comment by du82** [comment]
I feel like this is undesireable haha

https://github.com/status-im/status-mobile/assets/65834069/8e1812a4-a233-4983-86fa-6c217f007c2a

**Comment by vkjr** [comment]
@du82, thanks a lot! Wow, that looks like no change at all.. whereas on my android studio emulator and on genymotion emulator it behaves normally. Is it 100% that you downloaded the today's android build?


https://github.com/status-im/status-mobile/assets/5786310/df96e0a5-fc01-4afd-a51e-97d20d6304c4

**Comment by du82** [comment]
@vkjr I downloaded the latest PR from this build for the testing I did in the video

**Comment by pavloburykh** [comment]
> @pavloburykh, could you please check the cursor fix on Android? I don't have a physical device, but on the Simulator it works better)

Hi @vkjr! Confirming that issue is still reproducible on my Samsung Galaxy A52, Android 12. 

Seems like issue has become even more visible. 


https://github.com/status-im/status-mobile/assets/97245802/b7107a7b-b4d8-44fe-8c4c-e03e0702442e

**Comment by vkjr** [comment]
@pavloburykh, unbelievable... 🌚 
Thanks for the video, looks like I'm going to get a real device today)

**Comment by vkjr** [comment]
@du82, what is the Android version on your phone?
@pavloburykh can reproduce the issue on his Galaxy A52, Android 12. Whereas I can't reproduce it on physical Galaxy A23 with Android 13. And can't on Genymotion Pixel emulator with Android 12.

**Comment by VolodLytvynenko** [comment]
@vkjr similar issue to https://github.com/status-im/status-mobile/pull/16330#issuecomment-1611270810 happens on Pixel 7a, Android 13. 

In my case, cursor jumping to beginning of composer when reopening chat and continuing typing

### Steps to reproduce:
1. Open a chat.
2. Type a message in the composer but do not send it.
3. Close the chat.
4. Reopen the chat.
5. Start typing something in the composer again.

### Actual result:
When reopening the chat and continuing typing in the composer, the cursor jumps to the beginning of the composer input field.


https://github.com/status-im/status-mobile/assets/52490791/56696551-4db8-4fd5-b081-ae8564264198



### Expected result:
When reopening the chat and continuing typing in the composer, the cursor should remain at the current position

**Comment by pavloburykh** [comment]
@vkjr also, there is another issue related to highlighting (thanks @VladimrLitvinenko for finding)

### ISSUE 2 Sometimes mentions are not highlighted or text is highlighted instead of mentions

Reproducible both on IOS (14.1 iPhone X) and Android (12, Samsung Falaxy A52)

**Steps:** 

1. Expand the composer
2. Strat typing some multiline text
3. Put mentions in different parts of multiline text
4. Observe the result

**Actual result:** some mentions are not highlighted, sometimes message text is highlighted instead of mention


https://github.com/status-im/status-mobile/assets/97245802/13612e9b-476c-442a-a815-7e302b0ac207

photo_2023-07-07 16 53 06
photo_2023-07-07 16 53 03

**Comment by VladimrLitvinenko** [comment]
## ISSUE 2: Incorrect highlighting after leaving mentions in different places of composer

### Steps to reproduce:
1. Open the composer.
2. Type a message 
3. Leave first mention in the end of text
4. Tap in the beginning of composer and leave mention in the beginning

### Actual result:
The highlighting may occur for the piece of text instead of the mentioned user.
https://github.com/status-im/status-mobile/assets/52490791/23e38350-077d-440a-9011-525209142b91


### Expected result:
The highlighting occurs for mentioned user

### Device:
- Pixel 7a, Android 13

**Comment by vkjr** [comment]
@pavloburykh, I've committed a fix that removes cursor jumping to the beginning after reopening chat. It works on Android 13, but not sure if that fixes Android 12 behavior. Would you mind to check?

The cause of the problem is a [bug in react-native](https://github.com/facebook/react-native/issues/34695). The bug's essence is that if the developer sets the cursor position from the code using the `setNativeProps` method, the Android cursor is stuck on that position. And we did exactly this - setting initial cursor position when composer loaded.

**Comment by du82** [comment]
> @du82, what is the Android version on your phone? @pavloburykh can reproduce the issue on his Galaxy A52, Android 12. Whereas I can't reproduce it on physical Galaxy A23 with Android 13. And can't on Genymotion Pixel emulator with Android 12.

OnePlus Nord N20 5G (Chinese edition) - Android 11

**Comment by vkjr** [comment]
@qfrank, I believe Issue 2 described above looks is related to the implementation on status-go side. 

I can reproduce it with the following steps:
- type "@" without selecting mention
- type  "@" and select user to mention
- move cursor to the beginning, type "@" and select user

Result: incorrect syntax because of incorrect data structure coming from `status-go`

Video:
https://github.com/status-im/status-mobile/assets/5786310/5cec5a99-05eb-4694-a9f2-5b5ca96f118a

Data structure in the end:
Screenshot 2023-07-10 at 16 43 44


Would you mind to take a look? Thanks!

**Comment by pavloburykh** [comment]
> @pavloburykh, I've committed a fix that removes cursor jumping to the beginning after reopening chat. It works on Android 13, but not sure if that fixes Android 12 behavior. Would you mind to check?
> 
> The cause of the problem is a [bug in react-native](https://github.com/facebook/react-native/issues/34695). The bug's essence is that if the developer sets the cursor position from the code using the `setNativeProps` method, the Android cursor is stuck on that position. And we did exactly this - setting initial cursor position when composer loaded.

Thank you for the fix @vkjr ! After quick test I confirm that behaviour is much better. Although cursor still jumps slightly during fast typing (can be observed on video below). But I believe we can improve this behaviour in a followup. Tomorrow I will proceed with testing and ask other QAs to check on their Android 13

Also, will wait for feedback from @qfrank regarding [ISSUE 2](https://github.com/status-im/status-mobile/pull/16330#issuecomment-1629229938). 


https://github.com/status-im/status-mobile/assets/97245802/ab8788f4-1719-4dbf-9f72-e28a0a832e77

**Comment by pavloburykh** [comment]
@vkjr there is one more issue. I am not sure if it is related to ISSUE 2 or not, so I will log it separately

### ISSUE 3 Mention is highlighted despite another text is following mention without space

**Steps:** 

1. Enter a mention into inout field 
2. Type same text that will follow mention without space between them
3. See if mention is highlighted

**Actual result:** mention is highlighted. Such mention is not resolved after sending. 

https://github.com/status-im/status-mobile/assets/97245802/a10ac8ec-5e20-4a61-b4cf-0153152e46c9

**Expected result:** mention should not be highlighted because in fact it is invalid.

**Comment by vkjr** [comment]
@pavloburykh, yes, Issue 3 is also related to data returned from `status-go`
cc @qfrank

**Comment by qfrank** [comment]
Sorry for the delay since taking time off, will take a look next Monday cc @pavloburykh @vkjr 🙂

**Comment by vkjr** [comment]
@pavloburykh,
@qfrank already made fixes for mention highlighting issues in [`status-go` PR](https://github.com/status-im/status-go/pull/3766).
This mobile PR already includes those fixes. Could you please check? (I tested, works for me)

**Comment by pavloburykh** [comment]
> @pavloburykh, @qfrank already made fixes for mention highlighting issues in [`status-go` PR](https://github.com/status-im/status-go/pull/3766). This mobile PR already includes those fixes. Could you please check? (I tested, works for me)

Hi @vkjr @qfrank ! Thank you for the fixes. Currently the whole QA team is busy with release testing. As soon as we finish with release will continue with this PR.

**Comment by pavloburykh** [comment]
> @pavloburykh, @qfrank already made fixes for mention highlighting issues in [`status-go` PR](https://github.com/status-im/status-go/pull/3766). This mobile PR already includes those fixes. Could you please check? (I tested, works for me)

Hi @vkjr @qfrank ! Unfortunately [ISSUE 2](https://github.com/status-im/status-mobile/pull/16330#issuecomment-1625459510) is still reproducible for me

photo_2023-07-19 15 40 28

https://github.com/status-im/status-mobile/assets/97245802/e840cd5a-a0ed-48ad-9b0d-7698c5bd8a1c

**Comment by qfrank** [comment]
> > @pavloburykh, @qfrank already made fixes for mention highlighting issues in [`status-go` PR](https://github.com/status-im/status-go/pull/3766). This mobile PR already includes those fixes. Could you please check? (I tested, works for me)
> 
> Hi @vkjr @qfrank ! Unfortunately [ISSUE 2](https://github.com/status-im/status-mobile/pull/16330#issuecomment-1625459510) is still reproducible for me
> 
> photo_2023-07-19 15 40 28
> 
>  telegram-cloud-document-2-5314594743978570339.mp4

Hi @pavloburykh , after set log level to `debug`, can you share your logs, thank you

**Comment by pavloburykh** [comment]
> Hi @pavloburykh , after set log level to `debug`, can you share your logs, thank you

sure @qfrank ! Here are the logs and correspondent video

[Status-debug-logs (41).zip](https://github.com/status-im/status-mobile/files/12103387/Status-debug-logs.41.zip)

https://github.com/status-im/status-mobile/assets/97245802/57f19d04-2bd8-46a8-ba39-9ffe79053571

**Comment by qfrank** [comment]
> > Hi @pavloburykh , after set log level to `debug`, can you share your logs, thank you
> 
> sure @qfrank ! Here are the logs and correspondent video
> 
> [Status-debug-logs (41).zip](https://github.com/status-im/status-mobile/files/12103387/Status-debug-logs.41.zip)
> 
>  telegram-cloud-document-2-5317014640222286365.mp4

Thank you for your logs, it make reproducing easier and reproduced on status-go side now, i'll take a deep look

**Comment by qfrank** [comment]
I've updated my [PR](https://github.com/status-im/status-go/pull/3766) for the status-go side, could you help update `status-go-version.json` to point it to the latest changed version? the latest update should fix issue 2, thank you @vkjr  cc @pavloburykh

**Comment by vkjr** [comment]
@qfrank, thank you! Updated)

**Comment by pavloburykh** [comment]
@vkjr @qfrank issue 2 is fixed - thank you! I would prefer to run e2e before merging the PR, but currently we are experiencing some infrastructure problems. I believe that by tomorrow those issues will be resolved and I will be able to run e2e. Will keep updated. Sorry for delay and thank you for your patience and work on this PR.

**Comment by vkjr** [comment]
@pavloburykh, thank you!

**Comment by pavloburykh** [comment]
@vkjr there is one more obvious issue that I have missed, but thanx to @qoqobolo who pointed it out

### ISSUE 3 Mentions font weight in composer differs from the design

**Actual result:**

photo_2023-07-20 19 25 19


**Expected result:** 

https://www.figma.com/file/wA8Epdki2OWa8Vr067PCNQ/Composer-for-Mobile?type=design&node-id=5530-358489&mode=design&t=w3TQILfGs4ZdfPjM-0

photo_2023-07-20 19 24 14

**Comment by pavloburykh** [comment]
@vkjr also, as was noticed by @qoqobolo the cursor is still slightly jumping during fast typing. This might be leftover from [issue 1](https://github.com/status-im/status-mobile/pull/16330#issuecomment-1611270810). Below is video recorded in slow mo (Android 13, Pixel 7a). @vkjr WDYT, is it possible to improve this behaviour? Comparing to develop branch cursor behaviour seems to get worse. 




https://github.com/status-im/status-mobile/assets/97245802/f20c1e61-dde5-48f7-92d4-6eeee48b6940

**Comment by vkjr** [comment]
@pavloburykh, @qoqobolo, thank you for your hawk eyes!)

ezgif-2-27266dbb2f

**Comment by vkjr** [comment]
@pavloburykh, to be honest, so far I see no quick and easy ways to improve cursor behavior. It is definitely slower than develop because the way it renders changed with the addition of mentions highlighting.
Delay happens because text layout recalculates each time user changes input and that recalculation happens in `status-go`, so we send updated text there and wait for recalculations. Maybe that recalculation can partially happen on `status-mobile` side to make things quicker. Wdyt, @qfrank ?

**Comment by qfrank** [comment]
> Maybe that recalculation can partially happen on status-mobile side to make things quicker

I don't think it's easy to make recalculation partially happen on status-mobile side, all calculation is bundled tightly, except you want move logic of mention back to status-mobile again... if you think there is a performance issue on status-go side i can try to improve it cc @flexsurfer 

my guess:
the cursor issue seems relate to async rpc call, e.g.
1. OnChangeText("abc") return "abc"
2. OnChangeText("abcd") return "abcd"
but the second result("abcd") may reach earlier than the first one("abc") or something like this

my solution: add a param named call-seq . i can create PR based on this PR and see if there is improvement, WDYT? 🙂 @vkjr

**Comment by flexsurfer** [comment]
at least input without mention should work fast, so we should check in react if there is a mention in the input, just by detecting "@" symbol, and then we could try also to improve text with mention but with lower priority

**Comment by vkjr** [comment]
@qfrank, I think cursor issue is not related to the wrong order of arrived results. In this case, we'd have the disappeared symbols and we already had this issue and it is fixed)

In fact, when the user types, react always draw the typed symbol in text-input and then re-renders text-input content again using structures returned from `status-go`. So sometimes, when typing is fast, we see the latest symbols typed by the user, but `status-go` responses come later, and textinput rerenders with the correct symbols order, but a bit later. So it makes the impression that the text was typed, untyped, and typed again.

So cursor movements are really about delays between user input and re-rendering when `status-go` results arrived.

We can try something similar to what @flexsurfer suggested. When the user makes changes and they aren't inside mention, `status-go` call can be omitted and text structure updated on `status-mobile` side. But when mentions are involved - we call `status-go`. And we can try this in a separate PR))

**Comment by vkjr** [comment]
@churik, font-weight fixed) Could you please check? I'd be super-happy to merge this one)

**Comment by churik** [comment]
@vkjr 
there is an issue that disrupts the input order in the composer completely, which can be fixed only with the app reopening (please, wait for designers to reply to the questions below before proceeding with fixing)
 
### ISSUE 3: cutting the part of mention leads to autocomplete of the composer with an unrelated value

**Steps:**
1) open 1-1 chat, start typing 
2) type `@`, select the user 
3) select part of mention, tap on `Cut`, and continue to type

**Expected result:**
can continue 

**Actual result:**

https://github.com/status-im/status-mobile/assets/4557972/7aa34f24-4d3b-4197-bbe7-bb9623b0a22a





**OS:** Android 13 (Google Pixel 7)

@Francesca-G 

I'm contemplating several issues that already have been mentioned:
1) PR still does not implement the feature according to design, so there is still `@` in the input field in the front of each mention

2) As mentioned [here](https://github.com/status-im/status-mobile/pull/16330#issuecomment-1644251790) it added more complexity to the message input field and therefore may affect performance.

3) it is a source of bugs like ISSUE3 (see above) and frankly if I put on the scale the weight of real benefit from this feature and potential risks it may cause, I'd prefer to live without it.
WDYT?

@vkjr @J-Son89 my apologies, I know that it is not the right timing to push back on the feature, but seeing more and more risks just I'm compelled to ask one more time - is it required?

cc @John-44

**Comment by vkjr** [comment]
@churik, thank you for finding all these tricky issues!
I agree that this PR touches a lot of things) But we don't have to merge it now, we can improve it until we are fully happy with the result)

**Comment by J-Son89** [comment]
> @vkjr there is an issue that disrupts the input order in the composer completely, which can be fixed only with the app reopening (please, wait for designers to reply to the questions below before proceeding with fixing)
> 
> ### ISSUE 3: cutting the part of mention leads to autocomplete of the composer with an unrelated value
> **Steps:**
> 
> 1. open 1-1 chat, start typing
> 2. type `@`, select the user
> 3. select part of mention, tap on `Cut`, and continue to type
> 
> **Expected result:** can continue
> 
> **Actual result:**
> 
> https://github.com/status-im/status-mobile/assets/4557972/0f8d7c35-a659-401d-88@Francesca-G 277193ac02f6
> 
> **OS:** Android 13 (Google Pixel 7)
> 
> @Francesca-G
> 
> I'm contemplating several issues that already have been mentioned:
> 
> 1. PR still does not implement the feature according to design, so there is still `@` in the input field in the front of each mention
> 2. As mentioned [here](https://github.com/status-im/status-mobile/pull/16330#issuecomment-1644251790) it added more complexity to the message input field and therefore may affect performance.
> 3. it is a source of bugs like ISSUE3 (see above) and frankly if I put on the scale the weight of real benefit from this feature and potential risks it may cause, I'd prefer to live without it.
>    WDYT?
> 
> @vkjr @J-Son89 my apologies, I know that it is not the right timing to push back on the feature, but seeing more and more risks just I'm compelled to ask one more time - is it required?
> 
> cc @John-44

I have no strong opinions on this. I would agree with you @churik, as imo it would be great to have the fundamentals of the chat working really smoothly etc first and then start adding in other feature. However I'm not so involved with the messenger side of the app and as such I'm not  aware of what the strategy for the messenger is.
@flexsurfer might have some input on this :)

**Comment by qfrank** [comment]
update: just merged one small PR into this PR, can you help check if the typing experience becomes better? cc @qoqobolo thank you

**Comment by qoqobolo** [comment]
> update: just merged one small PR into this PR, can you help check if the typing experience becomes better? cc @qoqobolo thank you

@qfrank Just checked the latest build https://github.com/status-im/status-mobile/commit/2a1d0d95622d7fab0d7235048506b476b0762d5e and can say that the typing experience definitely got better, thank you.

But there are two new issues:

### ISSUE 4: No mentions suggested when typing `@`

**Steps:**
1. Open chat/community channel
2. Type `@`

**Actual result:** no suggestions
Screenshot 2023-07-26 at 19 41 14

### ISSUE 5: Text stays in composer after the message is sent

**Steps:**
1. Open chat/community channel
2. Type some text
3. Hit `Send`


https://github.com/status-im/status-mobile/assets/67952253/82317d59-5f9f-42e0-bed0-c9e3f5a20e17

**Comment by churik** [comment]
@Francesca-G can we get a design opinion on the necessity of this feature?

**Comment by pedro-et** [comment]
Text cant stay in composer after being sent.

**Comment by churik** [comment]
@pedro-et I'm asking about https://github.com/status-im/status-mobile/pull/16330#issuecomment-1649555182 :)

**Comment by qfrank** [comment]
thank you for your feedback @qoqobolo , issue 4-5 can be easily resolved, but i found something new, i'll push the fix later cc @vkjr

**Comment by churik** [comment]
@qfrank @vkjr 

Here is the [context](https://discord.com/channels/624634427930312714/975775548007678033/1134392911526641694)

My apologies for postponing the feature, taking into amount all the bugs and regression, I'd descope it for now.
сс @cammellos

**Comment by vkjr** [comment]
@churik, thank you for bringing up the discussion with design team!
I think we anyway need to get back to it and solve all related issues, because, as far as I understand, in future we want to have different stylings in text edit field (like making text bold). So this PR is a ground for similar future changes)

**Comment by qfrank** [comment]
issue 4-5 should be resolved also the chance of cursor jump should be fair low now, but let's stop here ATM and maybe get back to it and continue in some time in the future :)

**Comment by J-Son89** [comment]
> @churik, thank you for bringing up the discussion with design team! I think we anyway need to get back to it and solve all related issues, because, as far as I understand, in future we want to have different stylings in text edit field (like making text bold). So this PR is a ground for similar future changes)

@vkjr, @churik - if the issue is to what you described just now, that we want to be able to handle multiple stylings - perhaps a way to go about this is to instead just work on this component in isolation as it is in the design system. 
https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=3271-65639&mode=design&t=vvVr4Fa8c3f7jqmh-4 (although no bold text here - perhaps need to be updated to include this example too?)

Currently almost none of the design system components related to messaging are in our quo2 component library (I believe there was some initial confusion about how to make these components separate to our state management system and that's why it was not done this way - this should be easy to do ).
Maybe we can correctly add this component to the design system and implement the styling etc correctly so that it will be (hopefully) straightforward then to hook this up when the chat is in a stable state to add in this feature?

File: src/status_im2/contexts/chat/composer/style.cljs
Comment (J-Son89) [inline src/status_im2/contexts/chat/composer/style.cljs:132]: I think this color might be wrong - it should be based on the chat color or the users custom color?
Screenshot 2023-07-28 at 12 13 29

**Comment by John-44** [comment]
> @vkjr there is an issue that disrupts the input order in the composer completely, which can be fixed only with the app reopening (please, wait for designers to reply to the questions below before proceeding with fixing)
> 
> ### ISSUE 3: cutting the part of mention leads to autocomplete of the composer with an unrelated value
> **Steps:**
> 
> 1. open 1-1 chat, start typing
> 2. type `@`, select the user
> 3. select part of mention, tap on `Cut`, and continue to type
> 
> **Expected result:** can continue
> 
> **Actual result:**
> 
>  FILE.2023-07-25.14.27.29.mp4 
> **OS:** Android 13 (Google Pixel 7)
> 
> @Francesca-G
> 
> I'm contemplating several issues that already have been mentioned:
> 
> 1. PR still does not implement the feature according to design, so there is still `@` in the input field in the front of each mention
> 2. As mentioned [here](https://github.com/status-im/status-mobile/pull/16330#issuecomment-1644251790) it added more complexity to the message input field and therefore may affect performance.
> 3. it is a source of bugs like ISSUE3 (see above) and frankly if I put on the scale the weight of real benefit from this feature and potential risks it may cause, I'd prefer to live without it.
>    WDYT?
> 
> @vkjr @J-Son89 my apologies, I know that it is not the right timing to push back on the feature, but seeing more and more risks just I'm compelled to ask one more time - is it required?
> 
> cc @John-44

Sorry @churik , only just seeing this comment now.

If it makes development easier I think it's fine if we don't support partial deletions of @ mentions.  What I mean by this is that once an @ mention (e.g. "@John_") has been entered/finalised in the composer, if the caret is positioned immediately to the right of the finalised @ mention and the user presses backspace then the whole @ mention should be deleted (not just the last letter in the @ mention).

In terms of selection behaviour, if it helps development I think it's also fine if finalised @ mentions are treated for the purposes of selection as a single block e.g. the whole of the @ mention counts as a single letter and it's not possible to select some but not all of letters in the @ mention.  

Making these two changes would make it impossible to delete just part of a *finalized* @ mention.

Do the above only if it helps with development.

I may have misunderstood something about the problem or the question being asked.  If I have, please clarify.

cc'ing @vkjr @J-Son89 @qfrank @pedro-et @qoqobolo

**Pull request closed without merging**
